x/virtualgroup/keeper/keeper.go

```diff
@@ -647,3 +647,283 @@ func (k Keeper) CompleteSwapOut(ctx sdk.Context, gvgFamilyID uint32, gvgIDs []ui
+		if successorSP.Id != swapInInfo.SuccessorSpId {
+			return types.ErrSwapInFailed.Wrapf("The SP(ID: %d) has not reserved the swap(swapInfo=%s)", successorSP.Id, swapInInfo.String())
+		}
+		if uint64(ctx.BlockTime().Unix()) >= swapInInfo.ExpirationTime {
+			return types.ErrSwapInFailed.Wrapf("reserved swap expired")
+		}
+
+		targetPrimarySP, found := k.spKeeper.GetStorageProvider(ctx, swapInInfo.TargetSpId)
+		if !found {
+			return sptypes.ErrStorageProviderNotFound.Wrapf("The storage provider(ID: %d) not found when complete swap in.", swapInInfo.TargetSpId)
+		}
+		if err := k.completeSwapInFamily(ctx, successorSP, targetPrimarySP, gvgFamilyID); err != nil {
+			return err
+		}
+		store.Delete(key)
+	} else {
+		key := types.GetSwapInGVGKey(gvgID)
+		bz := store.Get(key)
+		if bz == nil {
+			return types.ErrSwapInFailed.Wrapf("The swap info not found in blockchain.")
+		}
+		k.cdc.MustUnmarshal(bz, &swapInInfo)
+		if successorSP.Id != swapInInfo.SuccessorSpId {
+			return types.ErrSwapInFailed.Wrapf("The sp(ID: %d) has not reserved the swap for secondary SP(ID: %d)", successorSP.Id, swapInInfo.TargetSpId)
+		}
+		targetSecondarySP, found := k.spKeeper.GetStorageProvider(ctx, swapInInfo.TargetSpId)
+		if !found {
+			return sptypes.ErrStorageProviderNotFound.Wrapf("The storage provider(ID: %d) not found when complete swap in.", swapInInfo.TargetSpId)
+		}
+		if err := k.completeSwapInGVG(ctx, successorSP.Id, targetSecondarySP.Id, gvgID); err != nil {
+			return err
+		}
+		store.Delete(key)
+	}
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventCompleteSwapIn{
+		StorageProviderId:          successorSP.Id,
+		TargetStorageProviderId:    swapInInfo.TargetSpId,
+		GlobalVirtualGroupFamilyId: gvgFamilyID,
+		GlobalVirtualGroupId:       gvgID,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) completeSwapInFamily(ctx sdk.Context, successorSP, targetPrimarySP *sptypes.StorageProvider, familyID uint32) error {
+	family, found := k.GetGVGFamily(ctx, familyID)
+	if !found {
+		return types.ErrGVGFamilyNotExist
+	}
+	srcStat := k.MustGetGVGStatisticsWithinSP(ctx, targetPrimarySP.Id)
+	dstStat := k.GetOrCreateGVGStatisticsWithinSP(ctx, successorSP.Id)
+
+	gvgs := make([]*types.GlobalVirtualGroup, len(family.GlobalVirtualGroupIds))
+	for i, gvgID := range family.GlobalVirtualGroupIds {
+		gvg, found := k.GetGVG(ctx, gvgID)
+		if !found {
+			return types.ErrGVGNotExist
+		}
+		if gvg.PrimarySpId != targetPrimarySP.Id {
+			return types.ErrSwapInFailed.Wrapf(
+				"the primary id (%d) in global virtual group does not match the target primary sp id (%d)", gvg.PrimarySpId, targetPrimarySP.Id)
+		}
+		// swap deposit
+		if !gvg.TotalDeposit.IsZero() {
+			coins := sdk.NewCoins(sdk.NewCoin(k.DepositDenomForGVG(ctx), gvg.TotalDeposit))
+			err := k.bankKeeper.SendCoins(ctx, sdk.MustAccAddressFromHex(targetPrimarySP.FundingAddress), sdk.MustAccAddressFromHex(successorSP.FundingAddress), coins)
```

Comment:
Why transferring old sp's income to new sp? Not get it.

Reply:
fixed, should revert the `from` and `to` here.